glyphs-reader/src/font.rs

```diff
@@ -2898,14 +2910,91 @@ impl Font {
     pub fn load(glyphs_file: &path::Path) -> Result<Font, Error> {
         let mut font = Self::load_raw(glyphs_file)?;
 
+        // ensure that glyphs with components that have bracket layers
+        // also have bracket layers.
+        font.align_bracket_layers();
+
         // propagate anchors by default unless explicitly set to false
         if font.custom_parameters.propagate_anchors.unwrap_or(true) {
             font.propagate_all_anchors();
         }
         Ok(font)
     }
 
-    // load without propagating anchors
+    /// if a glyph has components that have alternate layers, copy the layer
+    /// locations into the glyph.
+    ///
+    /// This is required to correctly handle bracket glyphs.
+    // https://github.com/googlefonts/glyphsLib/blob/c4db6b981d/Lib/glyphsLib/builder/bracket_layers.py#L176
+    fn align_bracket_layers(&mut self) {
+        // python does this recursively but we'll presort instead
+        let todo = self.depth_sorted_composite_glyphs();
+        //let mut problematic_glyphs = HashMap::new();
+
+        for name in &todo {
+            let mut needed = HashMap::new();
+            let glyph = self.glyphs.get(name).unwrap();
+            let my_bracket_layers = glyph
+                .bracket_layers
+                .iter()
+                .map(|l| l.bracket_info(&self.axes))
+                .collect::<HashSet<_>>();
+            for layer in &glyph.layers {
+                for comp in layer.components() {
+                    let comp_glyph = self.glyphs.get(&comp.name).unwrap();
+
+                    let comp_bracket_layers = comp_glyph
+                        .bracket_layers
+                        .iter()
+                        .map(|l| l.bracket_info(&self.axes))
+                        .collect::<HashSet<_>>();
+                    if comp_bracket_layers != my_bracket_layers {
+                        let missing = comp_bracket_layers.difference(&my_bracket_layers);
+                        needed
+                            .entry(layer.layer_id.clone())
+                            .or_insert(HashSet::new())
+                            .extend(missing.cloned());
+                    }
+                }
+            }
+
+            if needed.is_empty() {
+                // all good, next glyph
+                continue;
+            }
```

Comment:
This is a bit long and seems to have several distinct tasks that would benefit from becoming fns.

Suggest hoisting out a fn so you can have loop body only process things that need work, e.g.

```rust
for (name, (master, missing_brackets)) in todo.iter().filter_map(...compute Option<master, missing_brackets>...) {
   ...do the fixup work...
```

Reply:
this is another case where I am directly transcribing the python, and trying to match the structure as closely as possible because it makes it much easier to identify possible transcription errors.... several times in this kind of code I've used the wrong glyph name as a key or failed to invert a signed number or similar, and I find it very valuable to be able to look at the two impls side-by-side and and line-by-line.